### DISCUSSION TITLE
View page items?

### DISCUSSION THREAD

**Discussion by Alex-Wenner-FHR** [h2oai/wave#1065] [Q&A]
Is there a way to view page items? It seems like q.page['page'].items would be a list by the way we reference it (```q.page['page'].items[2]```) but I print it out and just get a wave component. I have a lot of elements on one page and struggling to reference it by index. Any thoughts on how I can find exactly what index a specific component is?

**Comment by mturoci** [comment]
cc @lo5

**Reply by Alex-Wenner-FHR** [reply]
Any thoughts on this one? I am was able to find out the element but by doing ``` q.page['page'].items[2], q.page['page'].items[3], q.page['page'].items[4] ... ``` This was very suboptimal, would be nice to see all the elements on a page!

**Reply by mturoci** [reply]
See https://github.com/h2oai/wave/issues/148.

**Comment by lo5** [comment]
There's no way to view the items inside `q.page`. This is because `q.page` is not a real collection, but a temporary placeholder for capturing all the modifications you do to the page before you call `q.page.save()`. 

In several of the scenarios that Wave supports, it's possible that different scripts/apps write to different parts of the _same_ page, so supplying each `q.page` in each app instance with a consistent view of what's contained in the actual page is an expensive operation.

I would recommend maintaining a separate `dict` of items that you add to the page, i.e. instead of using `q.page['foo'] = card`, call a separate function `set_card(q, 'foo', card)`:

```py
my_cards = dict() # your card collection

def set_card(q, key, card):
    q.page[key] = card
    my_cards[key] = card
```

This way, you can iterate through `my_cards` instead of attempting to iterate through `q.page`.
